toolvoice/scripts/inference/cascade_asr/asr_cascade_gemini_confetti.py

```
@@ -0,0 +1,378 @@
+        tool_cell = _coerce(r.get("tool"))
+
+        res = None
+        last_err = None
+        for attempt in range(3):
+            try:
+                res = await run_row(client, question, context, tool_cell)
+                break
+            except Exception as e:
+                last_err = e
+                msg = str(e).lower()
+                if "not found" in msg or "model" in msg and "live" not in msg:
```

Ruff (0.16.1)

[warning] Parenthesize `a and b` expressions when chaining `and` and `or` together, to make the precedence clear

Parenthesize the `and` subexpression

(RUF021)